src/api/blobs/reader.rs

```diff
@@ -0,0 +1,327 @@
+use std::{
+    io::{self, ErrorKind, SeekFrom},
+    pin::Pin,
+    task::{Context, Poll},
+};
+
+use n0_future::StreamExt;
+
+use crate::api::{
+    blobs::{Blobs, ReaderOptions},
+    proto::ExportRangesItem,
+};
+
+/// A reader for blobs that implements `AsyncRead` and `AsyncSeek`.
+#[derive(Debug)]
+pub struct BlobReader {
+    blobs: Blobs,
+    options: ReaderOptions,
+    state: ReaderState,
+}
+
+#[derive(Default, derive_more::Debug)]
+enum ReaderState {
+    Idle {
+        position: u64,
+    },
+    Seeking {
+        position: u64,
+    },
+    Reading {
+        position: u64,
+        #[debug(skip)]
+        op: n0_future::boxed::BoxStream<ExportRangesItem>,
+    },
+    #[default]
+    Poisoned,
+}
+
+impl BlobReader {
+    pub fn new(blobs: Blobs, options: ReaderOptions) -> Self {
+        Self {
+            blobs,
+            options,
+            state: ReaderState::Idle { position: 0 },
+        }
+    }
+}
+
+impl tokio::io::AsyncRead for BlobReader {
+    fn poll_read(
+        self: Pin<&mut Self>,
+        cx: &mut Context<'_>,
+        buf: &mut tokio::io::ReadBuf<'_>,
+    ) -> Poll<io::Result<()>> {
+        let this = self.get_mut();
+        let mut position1 = None;
+        loop {
+            let guard = &mut this.state;
+            match std::mem::take(guard) {
```

Comment:
fyi, I had this pattern show up being extremly slow in another project :'( with the rust compiler allocating over and over again, making reads/writes 

you can see how I fixed/worked around it here: https://github.com/rpgp/rpgp/pull/579/files#diff-70038cda95dcb4bc54877b1a3eb9f111ed226310dad30909b6d3cb88af6cd2f5

not saying this is currently slow, just warning you this can be very slow

Reply:
That is unfortunate. I like this pattern.

I think it should be fine as of now. The state is just 32 bytes, so it is very unlikely to matter compared with all the async stuff going on. I could rewrite it to just match on the state in place, but then there is always the chance that you get in some weird state instead of poisoned if you forget to handle an odd error.